### DISCUSSION TITLE
default initial memory size: how does it work? what's the max size I can set?

### DISCUSSION THREAD

**Discussion by jtorjo** [dotnet/runtime#84520] [Q&A]
@pavelsavara First off, thanks for this PR!

My problem is I'm working on a huge app, and it can end up throwing out-of-memory exceptions on Windows OSs with low memory (8GB of RAM).

I did use this setting to set the initial memory as high as possible, since I would assume the out-of-memory could be because of a "grow" failure when the memory needs to increase.

Q1: How does the initial memory get allocated? Does it ask the OS for it, but not actually allocate it?
 I've been setting this to 2GB from the get-go, and while the reported Total memory in JS is 2GB, I don't see that in Chrome -- the Chrome memory will start at roughly 1GB increase slowly to 2GB as my app uses more and more memory.
I'm not sure if this is how things are supposed to work.

Q2: What's the max size I can set?
I tried to set 3GB, and I got an error that the max size is not set or something.

**Comment by pavelsavara** [comment]
Initial size and max size are not the same thing.

Q1) I guess browser could do some mapping between real memory and the address space and that's why you see only 1GB at start. I don't know those browser implementation details.
Q2) The Net7 wasm can't use more than 2BG of address space. 

The underlying emscripten runtime could now address 4GB and we have open issue for that https://github.com/dotnet/runtime/issues/58382 

I'm not sure we are able to switch to 64bit address space.

> throwing out-of-memory exceptions on Windows OSs

I guess that the browsers would not like that either.

**Reply by jtorjo** [reply]
@pavelsavara Thanks!
I assumed initial size + max size are different :)

Q1: Got it
Q2: I see.... About 64bit address space -- I know it's not available yet, but I'd love it if it were.

> I guess that the browsers would not like that either.

Yes, agreed, but the "linear memory" in webassembly complicates stuff immensely. 

Lets say I have 2GB free memory.
I would assume if lets say I start with 1GB (thus, 1GB free remaining), and that gets full, and then webassembly wants to grow to 1.2GB, it would throw out-of-memory.

That's even though it could initially allocate 2GB from the get-go.
